src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,270 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/api/Migrate.h"
+#include "include/rados/librados.hpp"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/ImageState.h"
+#include "librbd/Operations.h"
+#include "librbd/Utils.h"
+#include "librbd/internal.h"
+#include "librbd/io/ImageRequestWQ.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::Migrate: "
+
+namespace librbd {
+namespace api {
+
+template <typename I>
+int Migrate<I>::migrate(I *ictx, librados::IoCtx& dst_io_ctx,
+                        const char *dstname_, ImageOptions& opts,
+                        ProgressContext &prog_ctx) {
+  CephContext* cct = ictx->cct;
+
+  std::string dstname = (dstname_ != nullptr && *dstname_ != '\0') ?
+    dstname_ : ictx->name.c_str();
+
+  ldout(cct, 10) << "migrate " << ictx->md_ctx.get_pool_name() << "/"
+                 << ictx->name << " -> " << dst_io_ctx.get_pool_name()
+                 << "/" << dstname << ", opts=" << opts << dendl;
+
+  if (ictx->read_only) {
+    return -EROFS;
+  }
+
+  std::list<obj_watch_t> watchers;
+  int r = ictx->md_ctx.list_watchers(ictx->header_oid, &watchers);
+  if (watchers.size() > 1) {
+    // TODO: filter out mirror watchers
+    lderr(cct) << "image has watchers - not migrating" << dendl;
+    return -EBUSY;
+  }
+
+  uint64_t features;
+  {
+    RWLock::RLocker snap_locker(ictx->snap_lock);
+    if (ictx->snaps.size() > 0) {
+      // XXXMG: or rather check for children?
```

Comment:
Hmm -- snapshots and clones are definitely an issue. I think this becomes limited value if we cannot transparently migrate at least snapshots. That logic can at least be copied from rbd-mirror image sync, but that implies the need for a special copy-up logic that can deep-sync snapshot state.

In fact, thinking about such a scenario -- as complicated as that seems -- might actually be a win-win. If we could have the "local" mirrored image be registered as a sync sink to the "remote" primary mirrored image, the actual block sync is really just basically a flatten, and we don't need to pause journal replay for a full bootstrap sync since the sync "copy-up" logic would ensure the underlying "remote" image block was properly deeply copied to the non-primary before writing the journal event to the block.